src/db/migrations/migration2.js

```diff
@@ -0,0 +1,33 @@
+const { getLogger } = require('../../utils/logger');
+const { map } = require('lodash');
+
+async function addTypeField(db) {
+  try {
+    // await db.update({ type: { $exists: false }, token: 'QTUM' }, { $set: { type: 'BET' } }, { multi: true });
+    // await db.update({ type: { $exists: false }, token: 'QTUM' }, { $set: { type: 'BET' } }, { multi: true });
+    const votes = await db.Votes.find({ type: { $exists: false } });
+    map(votes, async (vote) => {
+      if (vote.token === 'QTUM') {
+        await db.Votes.update({ txid: vote.txid }, { $set: { type: 'BET' } });
+      } else if (vote.token === 'BOT') {
+        const oracle = await db.Oracles.find({ address: vote.oracleAddress });
+        if (oracle[0].token === 'QTUM') await db.Votes.update({ txid: vote.txid }, { $set: { type: 'RESULT_SET' } });
+        else if (oracle[0].token === 'BOT') await db.Votes.update({ txid: vote.txid }, { $set: { type: 'VOTE' } });
+      }
+    });
+  } catch (err) {
+    getLogger().error(`DB update Error: ${err.message}`);
+    throw Error(`DB update Error: ${err.message}`);
+  }
+}
+
+async function migration2(db) {
+  try {
+    await addTypeField(db);
+  } catch (err) {
+    getLogger().error(err.message);
+    throw err;
+  }
+}
+
+module.exports = migration2;
```

Comment:
lets add a test case for the migration:
1. create a test db
2. add 2 oracles, 1 centralized, 1 decentralized
3. add 3 votes, 1 for each type
4. run the migration on this test db.
5. validate votes now have the type field

just copy some data from the dev graphql db

Reply:
What I did to test was 
1. copy the `votes.db` into a backup, then directly run migration on `votes.db`.
2. I check and compare the original number of entries with the number of `type` occurs, which is the same. 
3. Then I check if the type is correct by searching `oracle address` in `oracles.db`, after checking about 10 entries, I concluded that the result is what the code instructed.

Do you want me to do the test you mention here?

Reply:
I was referring to actual unit tests. Although your's test does work, its better to have repeatable unit tests so that you have test coverage for that specific area. 

For example, the migration works right now, but you don't have any unit tests for it. Then later down the line some code changes that breaks the migration. Then anyone running the migration from that point forward will have a broken server. Thus having test coverage on the migration would pick this up when someone broke the migration code.

Can you add the test cases I mentioned above?

Reply:
Will do it!

Reply:
I tried to add test.
1. I create mock db as a file in mock folder
2. votes.db is named into votes.copy
3. every time when running the test, it first removes existing `mock/votes`, then copy `mock/votes.copy` into `mock/votes`, then run migration on `mock/votes`
4. I have `initDB` based on the mock db in my test
5. but the migration did not work, need to spend more time on searching for reason, after migration works, I plan to `streamRead` the result file to check the `type`

Reply:
@BernardHan When adding the entries into the tables do it the native way:
1. `initDB()` -> create new db files: `votes.test.db` & `oracles.test.db` -> loadDatabase -> etc
2. `DBHelper.insertOracle()` x2 - centralized and decentralized
3. `db.Votes.insert(vote)` x3 - one for each type
4. run migration test
5. tests pass
6. delete `votes.test.db` & `oracles.test.db` with `fs.unlink()`